language-plutus-core/test/Spec.hs

```diff
@@ -156,6 +159,13 @@ allTests plcFiles rwFiles typeFiles typeErrorFiles evalFiles = testGroup "all te
     , test_evaluation
     , test_normalizationCheck
     , Check.tests
+    -- NEAT tests
+    , testCase "kind checker for generated types is sound" $
+        testTyProp depth kind prop_checkKindSound
+    , testCase "normalization preserves kinds" $
+        testTyProp depth kind prop_normalizePreservesKind
+    , testCase "normalization for generated types is sound" $
+        testTyProp depth kind prop_normalizeTypeSound
```

Comment:
Why do we hardcode `kind` in all of these cases? This way basically any type that starts with a lambda is going to be filtered out. Why do we need to pass any kind at all? Plutus Core is a fully elaborated language, we can infer the kind of any well-kinded type, so we don't need to try to guess the kind of a generated type.

The type checker should comprise two functions: `infer` and `check` (which is what we have in [`Typecheck.Internal`](https://github.com/input-output-hk/plutus/blob/05574c9c55eb06b26671b2d569625d7240ae747d/language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs)) where `infer` recurses on the type and `check` defers to `infer` and ensures that the inferred type is equal to the expected one. This way we'll be able to test stuff for types that are not of kind `*`.

Reply:
@effectfully The kind checker should be as eager as possible, able to reject partial types as soon as they become ill-typed. Therefore, splitting it into `infer` and `check` isn't possible.

Reply:
> The kind checker should be as eager as possible, able to reject partial types as soon as they become ill-typed. Therefore, splitting it into infer and check isn't possible.

We can reject partial types with `infer` as well, I guess it's just the case that a `Maybe Type` does not fit well into the API of `lazy-search`. I think we can probably provide a `Maybe a` instance for

```haskell
class Coolean b where
  toCool :: b -> Cool
  toBool :: b -> Bool
  isCool :: (a -> b) -> Bool
```

but it's not going to be as efficient as using the native `Cool`. Ok then, that:

> It's probably a good idea to switch these over to generating kinds, and then generating types for those kinds. However, most interesting types will probably be of the * kind, so that's what I chose for my initial set of tests.

makes sense to me.

Reply:
> > The kind checker should be as eager as possible, able to reject partial types as soon as they become ill-typed. Therefore, splitting it into infer and check isn't possible.
> 
> We can reject partial types with `infer` as well, I guess it's just the case that a `Maybe Type` does not fit well into the API of `lazy-search`. I think we can probably provide a `Maybe a` instance for
> 
> ```haskell
> class Coolean b where
>   toCool :: b -> Cool
>   toBool :: b -> Bool
>   isCool :: (a -> b) -> Bool
> ```
> 
> but it's not going to be as efficient as using the native `Cool`. Ok then, that:

The problem is the eagerness. We need to be able to reject the type while evaluation as little as possible. If you can make this work with an `infer`/`check` structure, have at it.

Reply:
> The problem is the eagerness. We need to be able to reject the type while evaluation as little as possible. 

I don't know if it's possible with `lazy-search`, but I'd expect it to be possible in theory. I mean, with `lazy-search` we're trying to fit type checking into something that returns essentially a boolean. But

```
infer :: TypeG n -> Maybe Kind
infer (Fun a b) = if check Star a && check Star b then Just Star else Nothing
infer (TyLam kind body) = KindArrow kind <$> infer body
<...>
```

is also lazy, can be parallelized, etc. It's just that `lazy-search` does not operate on `Maybe`, so we can't fit that laziness and parallelization into the provided API.

Anyway, your strategy of generating kinds first makes perfect sense to me (as well as not doing that in this PR).

Reply:
Hm, if we make it

```haskell
infer :: TypeG n -> Kind
infer (Fun a b) = if check Star a && check Star b then Star else error "wrong"
infer (TyLam kind body) = KindArrow kind $ infer body

Reply:
Hm, if we make it

```haskell
infer :: TypeG n -> Kind
infer (Fun a b) = if check Star a && check Star b then Star else error "wrong"
infer (TyLam kind body) = KindArrow kind $ infer body
```

and implement `check` in terms of `infer` and implement one additional function:

```haskell
kindcheck :: TypeG n -> Cool
kindcheck a = isRight . unsafePerformIO . try . evaluate . infer
```

then we can have a less awkward AST and a more direct type checker. I'm going to play with that at some point (this whole thing got me really curious, thanks for introducing me to it).

Reply:
I feel you're on your way to reimplementing NEAT.